File: frontend/src/app/services/hall/hall.service.ts
import { Injectable } from '@angular/core';
import { Globals } from '../../global/globals';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Hall} from '../../dtos/hall';
import {HallRequestParameter} from '../../datatype/requestParameters/HallRequestParameter';

@Injectable({
  providedIn: 'root'
})
export class HallService {

  private hallBaseUri: string = this.globals.backendUri + '/halls';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }


  /**
   * Loads all halls from backend
   */
  getAllHalls(): Observable<Hall[]> {
    return this.httpClient.get<Hall[]>(this.hallBaseUri);
  }

  /**
   * persists hall to the backend
   * @param hall to be added to the backend
   */
  createHall(hall: Hall) {
    console.log('Add hall to backend ' + hall.toString());
    return this.httpClient.post<Hall>(this.hallBaseUri, hall);
  }

  /**
   * searches for and loads hall from backend
   * @param hallName search parameter. If set only halls with names containing this parameter are found
   * @param hallLocation search parameter. If set only halls with this location are found
   * @param fields contain special instructions on which fields are to be included with found halls
   * For example if fields is set to id, only hall ids are found and returned from backend
   */
  searchHalls(hallName: string, hallLocation: Location, fields: HallRequestParameter[]): Observable<Hall[]> {
    console.log('Getting all halls from search parameters: name = ' + hallName + ', location = ' + JSON.stringify(hallLocation) +
      ' with ' + (fields ? ' fields ' + JSON.stringify(fields) : ' all fields'));
    let parameters = new HttpParams();
    parameters = fields ? parameters.append('fields', fields.toString()) : parameters;
    parameters = hallName ? parameters.append('name', hallName) : parameters;
    return this.httpClient.get<Hall[]>(this.hallBaseUri, {params: parameters});
  }

  /**
   * finds and loads a single hall entity by its id
   * @param id of hall to be found
   * @param include includes special request parameters
   * if set to 'editing' hall entities will include boolean declaring whether edits are allowed
   */
  findOneById(id: number, include: HallRequestParameter[]): Observable<Hall> {
    console.log('Get hall with id ' + id + ' and request parameters ' + include.toString());
    const parameters = include ? new HttpParams().set('include', include.toString()) : null;
    return this.httpClient.get<Hall>(this.hallBaseUri + '/' + id, {params: parameters});
  }

  /**
   * updates hall passed as parameter
   * @param hall entity to be updated with new parameters
   */
  updateHall(hall: Hall): Observable<Hall> {
    console.log('Update hall with id ' + hall.id + ' to ' + hall.toString());
    return this.httpClient.put<Hall>(this.hallBaseUri, hall);
  }

  /**
   * deletes hall with given id
   * @param id of hall to be deleted
   */
  deleteHall(id: number):  Observable<{}>  {
    console.log('Deleting hall with id ' + id);
    return this.httpClient.delete(this.hallBaseUri + '/' + id);
  }
}

File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/HallService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.hall.HallDTO;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.requestparameter.HallRequestParameter;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.searchParameters.HallSearchParametersDTO;
import at.ac.tuwien.sepm.groupphase.backend.exception.CustomValidationException;
import at.ac.tuwien.sepm.groupphase.backend.exception.ServiceException;

import java.util.List;

public interface HallService {

    /**
     * gets a list of halls found in the system filtered by search parameters if given
     * @param fields representing fields of hall entity that will be returned.
     * If null will return entire entity
     * @param searchParameters for which halls will be returned
     * if null will return all halls
     * @return list of all halls
     */
    List<HallDTO> findHalls(List<HallRequestParameter> fields, HallSearchParametersDTO searchParameters);

    /**
     * Adds hall to the system
     * @param hallDTO dto of entity to be added
     * @return created hall als dto
     * @throws ServiceException if something goes wrong during data processing
     * @throws CustomValidationException if hall already exists in the system
     */
    HallDTO addHall(HallDTO hallDTO) throws ServiceException, CustomValidationException;

    /**
     * finds specific hall by id
     * @param hallId id of hall to be found
     * @param include inclues special instructions for method.
     * If editing request parameter is set hall will return a boolean to decide hall should
     * be allowed to edit in frontend
     * @return hall with specified id
     */
    HallDTO findHallById(Long hallId, List<HallRequestParameter> include);

    /**
     * updates the passed as parameter
     * @param hallDTO to update hall to
     * @return updated hall as dto
     * @throws CustomValidationException if hall cannot be edited anymore because shows are associated with it
     */
    HallDTO updateHall(HallDTO hallDTO) throws CustomValidationException;

    /**
     * deletes the hall with the given id
     * @param hallId of hall to be deleted
     */
    void deleteHall(Long hallId);
}
